fvm/environment/unsafe_random_generator.go

```diff
@@ -11,34 +12,49 @@ import (
 
 type UnsafeRandomGenerator struct {
 	tracer *Tracer
-	rng    *rand.Rand
+
+	blockHeader *flow.Header
+
+	rng      *rand.Rand
+	seedOnce sync.Once
 }
 
 func NewUnsafeRandomGenerator(
 	tracer *Tracer,
 	blockHeader *flow.Header,
 ) *UnsafeRandomGenerator {
 	gen := &UnsafeRandomGenerator{
-		tracer: tracer,
+		tracer:      tracer,
+		blockHeader: blockHeader,
 	}
 
-	if blockHeader != nil {
+	return gen
+}
+
+// seed seeds the random number generator with the block header ID.
+// This allows lazy seeding of the random number generator,
+// since not a lot of transactions/scripts use it and the time it takes to seed it is not negligible.
+func (gen *UnsafeRandomGenerator) seed() {
+	gen.seedOnce.Do(func() {
+		if gen.blockHeader == nil {
+			return
+		}
 		// Seed the random number generator with entropy created from the block
 		// header ID. The random number generator will be used by the
 		// UnsafeRandom function.
-		id := blockHeader.ID()
+		id := gen.blockHeader.ID()
 		source := rand.NewSource(int64(binary.BigEndian.Uint64(id[:])))
 		gen.rng = rand.New(source)
-	}
-
-	return gen
+	})
 }
 
 // UnsafeRandom returns a random uint64, where the process of random number
 // derivation is not cryptographically secure.
 func (gen *UnsafeRandomGenerator) UnsafeRandom() (uint64, error) {
 	defer gen.tracer.StartExtensiveTracingSpanFromRoot(trace.FVMEnvUnsafeRandom).End()
 
+	gen.seed()
```

Comment:
It seems we want `seed` to ensure that `gen.rng` is ready after `seed` is called. 

Does sync.Once guarantee that?

I think sync.Once only guarantee if two calls happen concurrently, only one of them will run certain code. Is it possible the following event happens:

1. process A calls UnsafeRandom, triggers gen.seed to generate seed
2. process B concurrently calls UnsafeRandom, since process A is already generating seed, it skip immediately, and continue, then it hits `gen.rng == nil` and got a OperationNotSupportedError

I think we might need to use a mutex lock to block the call and ensure rng has been set before unblocking the call.

Please add a test to verify concurrent safe here.

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/sync/once.go;l=55-61

> Do guarantees that when it returns, f has finished.

Reply:
> process B concurrently calls UnsafeRandom

unsafe random cannot be concurrently called, as `gen.rng.Read(buf)` is already not concurrency safe. Not sure if concurrency is needed her, since this is within one transaction.Another argument against concurrency is that random needs to be deterministic. if process A and B call unsafe random at the same time, the random number A and B get should not depend on which is first. I think if we wanted concurrency here, each process would have their own instance of a deterministically pre-seeded random.